ci/generate_docs.py

```diff
@@ -0,0 +1,219 @@
+"""
+Automatic generator of documentation about CI jobs.
+Analyzes all .yml files connected with CI, takes the most important information
+and writes it into a README file.
+
+Features:
+- reads a job description from a comment above job definition
+- includes a link to each file and also to job definition
+- includes a script definition in a code block
+
+Usage:
+- put comments (starting with "#") directly above the job definition in .yml file
+
+Running the script:
+- `python generate_docs.py` to generate the documentation
+- `python generate_docs.py --test` to check if documentation is up-to-date
+"""
+
+import argparse
+import filecmp
+import os
+import re
+import sys
+from collections import OrderedDict
+from pathlib import Path
+from typing import Any, Dict, List
+
+import yaml
+
+# TODO: could read sections inside the file with some overall description of below jobs
+# TODO: could add something like "Generated automatically, do not edit by hand" into the README
+# TODO: when CI jobs are calling some make target, we could make a comment in
+#   that Makefile and take the text from there (so that Makefiles are documented as well)
+
+parser = argparse.ArgumentParser()
+parser.add_argument(
+    "--test",
+    action="store_true",
+    help="Check if there are no new changes in all CI .yml files",
+)
+args = parser.parse_args()
+
+
+class DocsGenerator:
+    def __init__(self) -> None:
+        # Going to the root directory, so the relative
+        # locations of CI files are valid
+        os.chdir(Path(__file__).resolve().parent.parent)
+
+        self.GITLAB_CI_FILE = ".gitlab-ci.yml"
+        self.DOC_FILE = "docs/ci/jobs.md"
+
+        # Some keywords that are not job definitions and we should not care about them
+        self.NOT_JOBS = [
+            "variables:",
+            "image:",
+            ".gitlab_caching:",
+        ]
+
+        self.ALL_JOBS: Dict[str, Dict[str, Any]] = OrderedDict()
+
+        self.FILES = self.get_all_ci_files()
+
+    def generate_docs(self) -> None:
+        """Whole pipeline of getting and saving the CI information."""
+        for file in self.FILES:
+            self.ALL_JOBS[file] = {
+                "jobs": self.get_jobs_from_file(file),
+                "overall_description": self.get_overall_description_from_file(file),
+            }
+
+        self.save_docs_into_file()
+
+    def verify_docs(self) -> None:
+        """Checking if the docs are up-to-date with current CI .yml files.
+
+        Creating a new doc file and comparing it against already existing one.
+        Exit with non-zero exit code when these files do not match.
+        """
+        already_filled_doc_file = self.DOC_FILE
+        self.DOC_FILE = "new_file_temp.md"
```

Comment:
Consider using [tempfile.TemporaryFile](https://docs.python.org/3/library/tempfile.html?highlight=mktemp#tempfile.TemporaryFile) which provides context manager for deletion.